### PR TITLE
ARROW-12028 ARROW-11940: [Rust][DataFusion] Add TimestampMillisecond support to GROUP BY/hash aggregates

### PR DESCRIPTION
**Pull request by velvia**
This PR adds support for TimestampMillisecondArray to `Scalar`, `GroupByScalar`, and hash_aggregate / GROUP BYs to Rust DataFusion, thus fixing 12028.   I believe it might also fix 11940, though this needs input from you all.

There are some formatting changes from running `cargo fmt`.

### CODE REVIEW

**Comment by velvia** [comment]
@nevilledips @andygrove

**Comment by alamb** [comment]
Thanks @velvia  -- I plan to review this tomorrow

File: rust/datafusion/src/scalar.rs [hunk 1]
```diff
@@ -19,6 +19,11 @@
 
 use std::{convert::TryFrom, fmt, iter::repeat, sync::Arc};
 
+use arrow::array::{
```
Comment (Dandandan) [inline rust/datafusion/src/scalar.rs:22]: These imports don't seem to be necessary? The module already has `array::*,` ?

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 2]
```diff
@@ -656,42 +656,20 @@ fn equal_rows(
...
-            DataType::UInt16 => {
-                equal_rows_elem!(UInt16Array, l, r, left, right)
-            }
-            DataType::UInt32 => {
-                equal_rows_elem!(UInt32Array, l, r, left, right)
-            }
-            DataType::UInt64 => {
-                equal_rows_elem!(UInt64Array, l, r, left, right)
-            }
+            DataType::Boolean => equal_rows_elem!(BooleanArray, l, r, left, right),
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:659]: Any idea why the formatting changed? Do you use a recent/stable rust/cargo/rustfmt version?

**Comment by Dandandan** [comment]
This looks good @velvia , I added some minor comments.

File: rust/datafusion/src/scalar.rs (hunk 1 above)
Comment (velvia) [reply rust/datafusion/src/scalar.rs:22]: Oh they were reformatted by `cargo fmt` which is what the README told me to run.....

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (velvia) [reply rust/datafusion/src/physical_plan/hash_join.rs:659]: This is what I'm running, is it too old?
`nightly-2020-11-25-x86_64-apple-darwin (default)`

Should arrow project have a `rust-toolchain` to specify what version of Rust to run on?

File: rust/datafusion/src/physical_plan/group_scalar.rs [hunk 3]
```diff
@@ -63,6 +64,7 @@ impl TryFrom<&ScalarValue> for GroupByScalar {
             ScalarValue::UInt16(Some(v)) => GroupByScalar::UInt16(*v),
             ScalarValue::UInt32(Some(v)) => GroupByScalar::UInt32(*v),
             ScalarValue::UInt64(Some(v)) => GroupByScalar::UInt64(*v),
+            ScalarValue::TimeMillisecond(Some(v)) => GroupByScalar::TimeMillisecond(*v),
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/group_scalar.rs:67]: Should we also add support for `ScalarValue::TimeMicrosecond` and `ScalarValue::TimeNanosecond`?

**Review by alamb — COMMENTED** [review]
Thanks @velvia  -- overall looks good. 

The only thing I think is needed for this PR is tests for grouping on columns with TimestampMillisecond.

File: rust/datafusion/src/physical_plan/group_scalar.rs (hunk 3 above)
Comment (velvia) [reply rust/datafusion/src/physical_plan/group_scalar.rs:67]: Sure thing, I wasn't sure why they weren't added here....

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (velvia) [reply rust/datafusion/src/physical_plan/hash_join.rs:659]: FWIW, I tried running cargo fmt with `+stable` on 1.50.0 and it did not change anything....

**Comment by velvia** [comment]
@alamb and others: have added a test and I believe addressed comments, hope this is all needed!

**Review by alamb — APPROVED** [review]
Looks good to me. Thanks @velvia !

**Comment by alamb** [comment]
@velvia  the `lint` job seems to have failed -- looks like the tests need to get formatted

https://github.com/apache/arrow/pull/9773/checks?check_run_id=2221026494

I tried to fix this PR by checking in the result of running `cargo fmt` locally,  but I don't have access to push to `urbanlogiq/arrow`

**Comment by velvia** [comment]
@alamb pushed the cargo fmt fix. :)

**Comment by velvia** [comment]
It seems the Docker build failed but I can't click on or expand on the details.

**Comment by alamb** [comment]
> It seems the Docker build failed but I can't click on or expand on the details.


Yeah, that has been happening recently. I don't think it is related to this PR. I plan to merge this one as soon as the remaining tests pass

**Pull request closed without merging**